### ISSUE TITLE
Flakes allow referring to non-existent store paths

### ISSUE THREAD

**Issue by dasJ** [NixOS/nix#5097]
**Describe the bug**

This is duplicating https://github.com/NixOS/nixpkgs/issues/132732 since I think it is a nix issue.
The issue was introduced with my Pull Request https://github.com/NixOS/nixpkgs/pull/131102

Essentially, when `environment.etc.?.source` refers to a path (`./zinputrc` in this case), it seems like the path the link is referring to is either not existant at all or it gets garbage collected.

**Steps To Reproduce**

These are alread described in https://github.com/NixOS/nixpkgs/issues/132732

**Expected behavior**

The symlink should be valid.

**`nix-env --version` output**

Not working: `nix-env (Nix) 2.4pre20210707_02dd6bb`

Working: `nix-env (Nix) 2.3.15`

**Additional context**

cc @Mic92

**Comment by waldheinz** [comment]
I can confirm this behavior, but it does not seem to be limited to flakes, for me nix-build also fails.  As I understand it, the `source` parameter is treated differently since eb7120dc79966d5ed168321fd213de38de13a2b1

Before that commit, the `source` parameter was converted to an string by implicit coercion, which would copy the file to the store and yield an string containing the store path. Now, by the virtue of `escapeShellArg`, `toString` is called explicitly on that path, which will yield an string containing the absolute path of the file.

We have entries like `environment.etc."mdi/logger.cfg".source = ./logger.cfg`, which result in dangling symlinks now. But the issue is also present in nixpkgs itself, e.g.  here:

> https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/bash/bash.nix#L199

This can be worked around, for us, by changing the entry to `environment.etc."mdi/logger.cfg".source = "${./logger.cfg}"`, restoring the "copy-to-store" behavior. But I'm not sure if this sufficient because there are potentially many places, also outside of  `nixpkgs`, relying on the old behavior.

I'll submit a PR restoring the old behavior, let's see what people think about that there.

**Comment by waldheinz** [comment]
Here's the PR: https://github.com/NixOS/nixpkgs/pull/136474

**Comment by erikarvstedt** [comment]
This is a bug in nixpkgs, not in nix. It was fixed in https://github.com/NixOS/nixpkgs/pull/136474. This issue can be closed.

**Comment by dasJ** [comment]
I agree this is fixed by the PR, but shouldn't nix prohibit these kinds of bugs in general?

**Comment by erikarvstedt** [comment]
What exactly is the bug that should be fixed? It's a valid use case of nix to include local paths in derivations.

**Comment by dasJ** [comment]
When the path is used as a string, it refers to a store path that is non-existent (of course, it cannot refer to the filesystem location since this is flakes after all).

The store path may be non-existent because it was wrong in the first place or maybe because it got garbage-collected, not really sure at this point.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
